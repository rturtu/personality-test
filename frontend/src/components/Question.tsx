import React, { useState } from "react";
import {
    PrimaryButton,
    SecondaryButton,
    VerticalRow,
} from "./StyledComponents";
import { Progress } from "semantic-ui-react";
import styled from "styled-components";
import { QuestionType, AnswerType } from "../types";

const QuestionWrapper = styled.div`
    &&& {
        width: 50vw;
        border: 1px solid ${(props) => props.theme.colors.lightBlue};
        padding: 2em;
        @media (max-width: ${(props) => props.theme.screenWidth.mobile}) {
            width: 90vw;
        }
        color: ${(props) => props.theme.colors.darkGrey};
    }
`;
const QuestionText = styled.p`
    &&& {
        font-size: 1.5em;
        align-text: left;
    }
`;
const AnswersWrapper = styled(VerticalRow)`
    &&& {
        text-align: left;
        font-size: 1.5em;
        & > * {
            margin-bottom: 0.7em;
        }
        margin-bottom: 1em;
    }
`;
const ButtonWrapper = styled.div`
    &&& {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        width: 100%;
        font-size: 1.3em;
    }
`;

interface Props {
    question: QuestionType;
    setAnswer: (questionId: number, answerId: number) => void;
    goBack: () => void;
    questionIndex: number;
    questionCount: number;
}

const QuestionComponent: React.FC<Props> = (props) => {
    const [selectedAnswerId, setSelectedAnswerId] = useState<number>(
        props.question.selectedAnswerId || -1
    );

    const isFirstQuestion = props.questionIndex === 0;
    const isLastQuestion = props.questionIndex === props.questionCount - 1;

    const sendAnswer = (): void => {
        if (selectedAnswerId === -1) return;
        props.setAnswer(props.question.id, selectedAnswerId);
    };

    return (
        <QuestionWrapper>
            {props.questionIndex >= 0 ? (
                <QuestionText>
                    Question {props.questionIndex + 1}
                    {props.questionCount && ` / ${props.questionCount}`}
                </QuestionText>
            ) : (
                <></>
            )}
            <QuestionText>{props.question.text}:</QuestionText>
            <AnswersWrapper>
                {props.question.answers.map((answer: AnswerType) => (
                    <div key={answer.id}>
                        <label onClick={() => setSelectedAnswerId(answer.id)}>
                            <input
                                type="radio"
                                checked={selectedAnswerId === answer.id}
                            />{" "}
                            <span>{answer.text}</span>
                        </label>
                    </div>
                ))}
            </AnswersWrapper>
            <ButtonWrapper>
                {isFirstQuestion ? (
                    <span></span>
                ) : (
                    <SecondaryButton onClick={props.goBack}>
                        Go Back
                    </SecondaryButton>
                )}
                <PrimaryButton onClick={sendAnswer}>
                    {isLastQuestion ? "See results" : "Next"}
                </PrimaryButton>
            </ButtonWrapper>
        </QuestionWrapper>
    );
};

export default QuestionComponent;
