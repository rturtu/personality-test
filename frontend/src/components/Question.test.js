import React from "react";
import Question from "./Question";
import renderer from "../helper/jest";

test("First question is displayed correctly", () => {
    const component = renderer(
        <Question
            question={{
                id: 1,
                text: "This is the first question",
                order: 1,
                answers: [
                    {
                        id: 1,
                        text: "The only answer",
                        order: 1,
                    },
                ],
            }}
            questionIndex={0}
            questionCount={4}
            setAnswer={(questionId, answerId) => {}}
            goBack={() => {}}
        />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Middle question is displayed correctly", () => {
    const component = renderer(
        <Question
            question={{
                id: 2,
                text: "This is some middle question",
                order: 2,
                answers: [
                    {
                        id: 1,
                        text: "The only answer",
                        order: 1,
                    },
                ],
            }}
            questionIndex={2}
            questionCount={4}
            setAnswer={(questionId, answerId) => {}}
            goBack={() => {}}
        />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Last question is displayed correctly", () => {
    const component = renderer(
        <Question
            question={{
                id: 4,
                text: "This is the last question",
                order: 4,
                answers: [
                    {
                        id: 2,
                        text: "The only answer",
                        order: 1,
                    },
                ],
            }}
            questionIndex={3}
            questionCount={4}
            setAnswer={(questionId, answerId) => {}}
            goBack={() => {}}
        />
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
