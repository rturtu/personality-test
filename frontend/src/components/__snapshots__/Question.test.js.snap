// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First question is displayed correctly 1`] = `
<div
  className="sc-iCoGMd sRtdd"
>
  <p
    className="sc-fujyAs jbcxtn"
  >
    Question 
    1
     / 4
  </p>
  <p
    className="sc-fujyAs jbcxtn"
  >
    This is the first question
    :
  </p>
  <div
    className="sc-gKAaRy sc-pNWdM fKPiYy fpXnMF"
  >
    <div>
      <label
        onClick={[Function]}
      >
        <input
          checked={false}
          type="radio"
        />
         
        <span>
          The only answer
        </span>
      </label>
    </div>
  </div>
  <div
    className="sc-jrsJWt qbsTD"
  >
    <span />
    <button
      className="ui button sc-bdnxRM sc-gtsrHT jtoVTB hxeQqD"
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;

exports[`Last question is displayed correctly 1`] = `
<div
  className="sc-iCoGMd sRtdd"
>
  <p
    className="sc-fujyAs jbcxtn"
  >
    Question 
    4
     / 4
  </p>
  <p
    className="sc-fujyAs jbcxtn"
  >
    This is the last question
    :
  </p>
  <div
    className="sc-gKAaRy sc-pNWdM fKPiYy fpXnMF"
  >
    <div>
      <label
        onClick={[Function]}
      >
        <input
          checked={false}
          type="radio"
        />
         
        <span>
          The only answer
        </span>
      </label>
    </div>
  </div>
  <div
    className="sc-jrsJWt qbsTD"
  >
    <button
      className="ui button sc-bdnxRM sc-dlnjwi jtoVTB gTPvDX"
      onClick={[Function]}
    >
      Go Back
    </button>
    <button
      className="ui button sc-bdnxRM sc-gtsrHT jtoVTB hxeQqD"
      onClick={[Function]}
    >
      See results
    </button>
  </div>
</div>
`;

exports[`Middle question is displayed correctly 1`] = `
<div
  className="sc-iCoGMd sRtdd"
>
  <p
    className="sc-fujyAs jbcxtn"
  >
    Question 
    3
     / 4
  </p>
  <p
    className="sc-fujyAs jbcxtn"
  >
    This is some middle question
    :
  </p>
  <div
    className="sc-gKAaRy sc-pNWdM fKPiYy fpXnMF"
  >
    <div>
      <label
        onClick={[Function]}
      >
        <input
          checked={false}
          type="radio"
        />
         
        <span>
          The only answer
        </span>
      </label>
    </div>
  </div>
  <div
    className="sc-jrsJWt qbsTD"
  >
    <button
      className="ui button sc-bdnxRM sc-dlnjwi jtoVTB gTPvDX"
      onClick={[Function]}
    >
      Go Back
    </button>
    <button
      className="ui button sc-bdnxRM sc-gtsrHT jtoVTB hxeQqD"
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
